pandas/io/parsers.py

```diff
@@ -1844,6 +1844,28 @@ def _do_date_conversions(self, names, data):
 
         return names, data
 
+    def _check_data_length(self, columns: List[str], data: List[np.ndarray]):
+        """Checks if length of data is equal to length of column names. One set of
+        trailing commas is allowed.
+
+        Parameters
+        ----------
+        columns: list of column names
+        data: list of array-likes containing the data column-wise
+
+        """
+        if not self.index_col and len(columns) != len(data) and columns:
+            if len(columns) == len(data) - 1 and np.all(
+                (data[-1] == "") | isna(data[-1])
```

Comment:
this seems a *very* specific condition. can you relax it?

Reply:
Do we want to allow more than one set of trailing commas? In this case we can remove the len check.
The array representing the last entries has either only nans or empty strings, this check is necessary

Reply:
don't we want to warn if there is a matchmatch at all?

Reply:
I thought we wanted to warn if we have more data-columns than names/headers except if we have trailing commas?

Reply:
do we have a test for that? I would warn regardless

Reply:
Changed this based on https://github.com/pandas-dev/pandas/pull/38587#issuecomment-750933280

test_no_header_two_extra_columns checks the warning and https://github.com/pandas-dev/pandas/blob/b5707d6b83d7476bc7fd0b3ccd30af0a7d6cba1a/pandas/tests/io/parser/test_common.py#L1066 checks trailing commas which does not raise a Warning